flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/ProcTimeInnerJoin.scala

```diff
@@ -0,0 +1,316 @@
+      out: Collector[CRow]): Unit = {
+
+    processElement(
+      valueC,
+      ctx,
+      out,
+      leftStreamWindowSize,
+      timerState1,
+      row1MapState,
+      row2MapState,
+      rightStreamWindowSize,
+      true
+    )
+  }
+
+  /**
+    * Process rightstream records
+    *
+    * @param valueC The input value.
+    * @param ctx   The ctx to register timer or get current time
+    * @param out   The collector for returning result values.
+    *
+    */
+  override def processElement2(
+      valueC: CRow,
+      ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+      out: Collector[CRow]): Unit = {
+
+    processElement(
+      valueC,
+      ctx,
+      out,
+      rightStreamWindowSize,
+      timerState2,
+      row2MapState,
+      row1MapState,
+      leftStreamWindowSize,
+      false
+    )
+  }
+
+  /**
+    * Called when a processing timer trigger.
+    * Expire left/right records which earlier than current time - windowsize.
+    *
+    * @param timestamp The timestamp of the firing timer.
+    * @param ctx       The ctx to register timer or get current time
+    * @param out       The collector for returning result values.
+    */
+  override def onTimer(
+      timestamp: Long,
+      ctx: CoProcessFunction[CRow, CRow, CRow]#OnTimerContext,
+      out: Collector[CRow]): Unit = {
+
+    if (timerState1.value == timestamp) {
+      expireOutTimeRow(
+        timestamp,
+        leftStreamWindowSize,
+        row1MapState,
+        timerState1,
+        ctx
+      )
+    }
+
+    if (timerState2.value == timestamp) {
+      expireOutTimeRow(
+        timestamp,
+        rightStreamWindowSize,
+        row2MapState,
+        timerState2,
+        ctx
+      )
+    }
+  }
+
+  /**
+    * Puts an element from the input stream into state and search the other state to
+    * output records meet the condition, and registers a timer for the current record
+    * if there is no timer at present.
+    */
+  private def processElement(
+      valueC: CRow,
+      ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+      out: Collector[CRow],
+      winSize: Long,
+      timerState: ValueState[Long],
+      rowMapState: MapState[Long, JList[Row]],
+      oppoRowMapState: MapState[Long, JList[Row]],
+      oppoWinSize: Long,
+      isLeft: Boolean): Unit = {
+
+    cRowWrapper.out = out
+    cRowWrapper.setChange(valueC.change)
+
+    val value = valueC.row
+
+    val curProcessTime = ctx.timerService.currentProcessingTime
+    val oppoExpiredTime =
+      if (oppoWinSize == 0) Long.MinValue else curProcessTime - oppoWinSize
+
+    // only when windowsize != 0, we need to store the element
+    if (winSize != 0) {
+      // register a timer to expire the element
+      if (timerState.value == 0 && winSize != -1) {
+        ctx.timerService.registerProcessingTimeTimer(curProcessTime + winSize + 1)
+        timerState.update(curProcessTime + winSize + 1)
+      }
+
+      var rowList = rowMapState.get(curProcessTime)
```

Comment:
The remove operation here is just an additional operation, it will expired in onTimer() anyway.
Yes, when winSize == 0, the record will not be put into state.